File: [0123][Hard][Best_Time_to_Buy_and_Sell_Stock_III]/Best_Time_to_Buy_and_Sell_Stock_III_3.py
# solution 2, 2 times traversal.
# leetcode time     cost : 92 ms
# leetcode memory   cost : 14.4 MB 
# Time  Complexity: O(N)
# Space Complexity: O(N)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        
        # forward traversal, profits record the max profit 
        # by the ith day, this is the first transaction
        profits = []
        max_profit = 0
        current_min = prices[0]
        for price in prices:
            current_min = min(current_min, price)
            max_profit = max(max_profit, price - current_min)
            profits.append(max_profit)
        
        # backward traversal, max_profit records the max profit
        # after the ith day, this is the second transaction 
        total_max = 0    
        max_profit = 0
        current_max = prices[-1]
        for i in range(len(prices) - 1, -1, -1):
            current_max = max(current_max, prices[i])
            max_profit = max(max_profit, current_max - prices[i])
            total_max = max(total_max, max_profit + profits[i])
            
        return total_max

File: [0123][Hard][Best_Time_to_Buy_and_Sell_Stock_III]/Best_Time_to_Buy_and_Sell_Stock_III_3.java
// solution 1, dp with dp compacted.
// leetcode time     cost : 1 ms
// leetcode memory   cost : 36.7 MB 
// Time  Complexity: O(kN)
// Space Complexity: O(k+N)
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int dp[] = new int[3];
        int cost[] = new int[3];
        Arrays.fill(cost, prices[0]);         //init all min elements with the value of the start price, price[0],
        for (int i = 1; i < prices.length; i++)  {
            for (int k = 1; k <= 2; k++) {
                cost[k] = Math.min(cost[k], prices[i] - dp[k-1]);
                dp[k] = Math.max(dp[k], prices[i] - cost[k]);
            }
        }
        return dp[2];
    }
}